CIAOS-Rust/server/src/storage.rs

```diff
@@ -0,0 +1,155 @@
+        self.file.read_exact(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+
+// this function accepts the flatbuffer and returns the offset and size list after proess the files in it
+pub fn write_files_to_storage(user : &str,body: &[u8])
+    -> Result<Vec<(u64, u64)>, Error> {
+    // Open the storage file "haystack.bin" in append mode
+    let mut haystack = OpenFile::new(&user)?;
+
+    let mut offset_size_list: Vec<(u64, u64)> = Vec::new();
+
+    //deserialises the binary data thats been accepted into FileDataList object using flatbuffer
+    
+    let file_data_list = match root::<FileDataList>(&body) {
+        Ok(data) => data,
+        Err(e) => return Err(ErrorBadRequest(format!("Failed to parse FlatBuffers data: {:?}", e))),
+    };
+    let files = match file_data_list.files() {
+        Some(files) => files,
+        None => return Err(ErrorBadRequest("No files found in FlatBuffers data")),
+    };
+    info!("Deserialized {} files", files.len());
+
+    for (index, file_data) in files.iter().enumerate() {
+        let data = match file_data.data() {
+            Some(data) => data,
+            None => {
+                error!("No data in file at index {}", index);
+                continue;
+            }
+        };
+
+        match haystack.write(data.bytes()) {
+            Ok((offset, size)) => {
+                offset_size_list.push((offset, size));
+                info!("Written file {} at offset {} with size {}", index, offset, size);
+            }
+            Err(e) => {
+                error!("Failed to write file {} to haystack: {}", index, e);
+                return Err(ErrorInternalServerError(e));
+            }
+        }
+    }
+
+    Ok(offset_size_list)
+}
+
+
+pub fn get_files_from_storage(user : &str, offset_size_list: Vec<(u64, u64)>)-> Result<Vec<u8>, Error> {
+    info!("connecting to .bin and gettting files");
+    let mut haystack = OpenFile::new(&user).map_err(ErrorInternalServerError)?;
+    info!("connected to .bin");
+    let mut builder = FlatBufferBuilder::new();
+    let mut file_data_vec = Vec::new();
+    info!("building the flatbuffer to share");
+    for &(offset, size) in offset_size_list.iter() {
+        let data = haystack.read(offset, size).map_err(ErrorInternalServerError)?;
+        let data_vector = builder.create_vector(&data);
+        let file_data = FileData::create(&mut builder, &FileDataArgs { data: Some(data_vector) });
+        file_data_vec.push(file_data);
+    }
+    info!("successfully built the buffer");
+    let files = builder.create_vector(&file_data_vec);
+    let file_data_list = FileDataList::create(&mut builder, &FileDataListArgs { files: Some(files) });
+    builder.finish(file_data_list, None);
+    info!("sending buffer");
+    Ok(builder.finished_data().to_vec())
+    }
+
+////////////////////////////////////////
+/// here starts delete functionality ///
+////////////////////////////////////////
+
+struct DeleteFile {
```

Comment:
why do we need two strcuts? one for open file and delete file?

Reply:
yes they both got different implementation